### PR TITLE
Correct all world height validations.

### PR DESCRIPTION
**Pull request by tommysanterre**
Addesses issue https://github.com/mc-server/MCServer/issues/1606.

No more assertion triggered.

### CODE REVIEW

File: src/BlockArea.cpp [hunk 1]
```diff
@@ -441,10 +441,10 @@ bool cBlockArea::Read(cForEachChunkProvider * a_ForEachChunkProvider, int a_MinB
 		LOGWARNING("%s: MaxBlockY less than zero, adjusting to zero", __FUNCTION__);
 		a_MaxBlockY = 0;
 	}
-	else if (a_MaxBlockY > cChunkDef::Height)
+	else if (a_MaxBlockY >= cChunkDef::Height)
 	{
 		LOGWARNING("%s: MaxBlockY more than chunk height, adjusting to chunk height", __FUNCTION__);
-		a_MaxBlockY = cChunkDef::Height;
+		a_MaxBlockY = cChunkDef::Height - 1;
```
Comment (madmaxoft) [inline src/BlockArea.cpp]: This breaks stuff. `a_MaxBlockY` is supposed to be "one block past the maximum" at this point (see lines 423 - 425), so with your change there's no way to read the top of the chunk

File: src/Generating/ChunkDesc.cpp [hunk 2]
```diff
@@ -336,10 +336,10 @@ void cChunkDesc::ReadBlockArea(cBlockArea & a_Dest, int a_MinRelX, int a_MaxRelX
 		LOGWARNING("%s: MaxRelY less than zero, adjusting to zero", __FUNCTION__);
 		a_MaxRelY = 0;
 	}
-	else if (a_MaxRelY > cChunkDef::Height)
+	else if (a_MaxRelY >= cChunkDef::Height)
 	{
 		LOGWARNING("%s: MaxRelY more than chunk height, adjusting to chunk height", __FUNCTION__);
-		a_MaxRelY = cChunkDef::Height;
+		a_MaxRelY = cChunkDef::Height - 1;
```
Comment (madmaxoft) [inline src/Generating/ChunkDesc.cpp]: Again, the max is supposed to be "one past", so the original test was correct.

File: src/Generating/FinishGen.cpp [hunk 3]
```diff
@@ -124,9 +124,9 @@ void cFinishGenNetherClumpFoliage::TryPlaceClump(cChunkDesc & a_ChunkDesc, int a
 	}
 
 	int MaxY = a_RelY + 2;
-	if (MaxY > cChunkDef::Height)  // Check if the coordinate is outside the chunk. If it it then adjust it.
+	if (MaxY >= cChunkDef::Height)  // Check if the coordinate is outside the chunk. If it it then adjust it.
 	{
-		MaxY = cChunkDef::Height;
+		MaxY = cChunkDef::Height - 1;
```
Comment (madmaxoft) [inline src/Generating/FinishGen.cpp]: The original code was correct, `MaxY` is used as an non-inclusive bound in the for loop below.

File: src/Generating/RoughRavines.cpp [hunk 4]
```diff
@@ -139,7 +139,7 @@ class cRoughRavine :
 			int NumBlocks = (rnd % 3) + 2;
 			rnd = rnd / 4;
 			float Val = (float)(rnd % 256) / 128 - 1;  // Random float in range [-1, +1]
-			if (h + NumBlocks > cChunkDef::Height)
+			if (h + NumBlocks >= cChunkDef::Height)
```
Comment (madmaxoft) [inline src/Generating/RoughRavines.cpp]: Wrong change.

File: src/World.cpp [hunk 5]
```diff
@@ -533,7 +533,7 @@ void cWorld::InitializeSpawn(void)
 void cWorld::Start(void)
 {
 	m_SpawnX = 0;
-	m_SpawnY = cChunkDef::Height;
+	m_SpawnY = cChunkDef::Height - 1;
```
Comment (madmaxoft) [inline src/World.cpp]: Nope. We don't want to spawn *in* the terrain, but above it.

File: src/BlockArea.cpp (hunk 1 above)
Comment (tommysanterre) [inline src/BlockArea.cpp]: Thank you, I didn't notice this. I could rename the variables to make it clearer we're working with intervals. Maybe using iterators like naming.
Example: 
    a_MinBlockX becomes a_BeginBlockX
    a_MaxBlockX becomes a_EndBlockX

File: src/Generating/ChunkDesc.cpp (hunk 2 above)
Comment (tommysanterre) [inline src/Generating/ChunkDesc.cpp]: I could rename the variable here as mentioned above.

File: src/Generating/FinishGen.cpp (hunk 3 above)
Comment (tommysanterre) [inline src/Generating/FinishGen.cpp]: And here as well.

File: src/BlockArea.cpp (hunk 1 above)
Comment (madmaxoft) [inline src/BlockArea.cpp]: No, I don't like that naming, it doesn't convey that the coords need to be sorted, and it indicates no inclusion either. Additionally, the original parameters are inclusive, but the maxima are made exclusive within the function, so either way you name them, one instance will be wrong. I prefer the interface part to be right.

Comment (tommysanterre) [inline src/BlockArea.cpp]: Oh I see that's the goal of that block. Make them exclusive.

	// Include the Max coords:
	a_MaxBlockX += 1;
	a_MaxBlockY += 1;
	a_MaxBlockZ += 1;

My proposition doesn't make sense then, sorry.

File: src/BlockArea.cpp [hunk 6]
```diff
@@ -441,7 +441,7 @@ bool cBlockArea::Read(cForEachChunkProvider * a_ForEachChunkProvider, int a_MinB
 		LOGWARNING("%s: MaxBlockY less than zero, adjusting to zero", __FUNCTION__);
 		a_MaxBlockY = 0;
 	}
-	else if (a_MaxBlockY > cChunkDef::Height)
+	else if (a_MaxBlockY >= cChunkDef::Height)
```
Comment (madmaxoft) [inline src/BlockArea.cpp]: I still think this is a wrong change, we need to be able to read areas with full chunk height (`a_MaxBlockY` is exclusive at this point due to line 425)

File: src/BlockArea.cpp [hunk 7]
```diff
@@ -517,7 +517,7 @@ bool cBlockArea::Write(cForEachChunkProvider * a_ForEachChunkProvider, int a_Min
 		LOGWARNING("%s: MinBlockY less than zero, adjusting to zero", __FUNCTION__);
 		a_MinBlockY = 0;
 	}
-	else if (a_MinBlockY > cChunkDef::Height - m_Size.y)
+	else if (a_MinBlockY >= cChunkDef::Height - m_Size.y)
```
Comment (madmaxoft) [inline src/BlockArea.cpp]: I'm not sure about this change either; if it is indeed needed, then the assignment on line 523 is wrong.

File: src/BlockArea.cpp (hunk 6 above)
Comment (tommysanterre) [inline src/BlockArea.cpp]: You're right.

File: src/BlockArea.cpp (hunk 7 above)
Comment (tommysanterre) [inline src/BlockArea.cpp]: Here I thought it was possible to call FastSetBlock with an invalid value but I can't reproduce it even when entering incorrect values with the debugger. I will revert this change also.

**Comment by tommysanterre** [comment]
@madmaxoft I really appreciate the reviews you have done for me. I've revised my changes once more and corrected some mistake I also fixed the issue highlighted by your last review. Could you review the code once more?

**Comment by madmaxoft** [comment]
Thanks for all the hard work :)

**Pull request merged**